### PR TITLE
vmap support for torch.linalg.vander

### PR DESCRIPTION
**Pull request by jazzysoggy**
Adds vmap support for torch.linalg.vander in a similar manner to how view_as_complex is implemented.

#91700

cc @zou3519 @Chillee @samdow @soumith @kshitij12345 @janeyx99

### CODE REVIEW

**Comment by jazzysoggy** [comment]
@kshitij12345 How do I fix the failure due to the test passing instead of failing?

**Comment by kshitij12345** [comment]
@jazzysoggy 

From the CI failure, we can see that `test_unimplemented_batched_registration` is failing for `linalg_vander`. So we can assume that we are marking this op as expected failure for this test. However, since this PR fixes that (by adding a batching rule), we will have to remove that xfail.

I would suggest to `grep` for `test_unimplemented_batched_registration` and try to find the relevant xfails for it.

Let me know if you need more suggestions :)

CI Failure Snippet:
```
======================================================================
[2226](https://github.com/pytorch/pytorch/actions/runs/3850590625/jobs/6561243328#step:10:2228)
XPASS [0.001s]: test_unimplemented_batched_registrations_[aten::linalg_vander] (__main__.TestFunctorchDispatcher)
[2227](https://github.com/pytorch/pytorch/actions/runs/3850590625/jobs/6561243328#step:10:2229)
----------------------------------------------------------------------
```

File: aten/src/ATen/functorch/BatchRulesDecompositions.cpp [hunk 1]
```diff
@@ -144,7 +144,7 @@ TORCH_LIBRARY_IMPL(aten, FuncTorchBatchedDecomposition, m) {
   OP_DECOMPOSE2(less, Tensor );
   OP_DECOMPOSE(linalg_cond);
   OP_DECOMPOSE(linalg_cholesky);
-  OP_DECOMPOSE(linalg_det);
+  OP_DECOMPOSE(c);
```
Comment (lezcano) [inline aten/src/ATen/functorch/BatchRulesDecompositions.cpp:147]: This change looks unintended.

File: aten/src/ATen/functorch/BatchRulesDecompositions.cpp [hunk 2]
```diff
@@ -159,6 +159,8 @@ TORCH_LIBRARY_IMPL(aten, FuncTorchBatchedDecomposition, m) {
   OP_DECOMPOSE(linalg_svd);
   OP_DECOMPOSE(linalg_svdvals);
   OP_DECOMPOSE(linalg_tensorinv);
+  OP_DECOMPOSE(linalg_vander);
+  OP_DECOMPOSE(linalg_vander, int);
```
Comment (lezcano) [inline aten/src/ATen/functorch/BatchRulesDecompositions.cpp:163]: I think this line is incorrect and unnecessary.

**Review by lezcano — CHANGES_REQUESTED** [review]

**Review by lezcano — COMMENTED** [review]
https://github.com/pytorch/pytorch/pull/91749?notification_referrer_id=NT_kwDOADI4gbI1MjQ2NzgyNTkwOjMyOTEyNjU#issuecomment-1374601689 is still missing, but otherwise it should be good.

Consider running locally the tests in `test/functorch` for `linalg_vander` to avoid running through too many CI cycles.

**Comment by kshitij12345** [comment]
@jazzysoggy It seems that you will also have to remove the following xfail as this PR implements batching rule for `linalg_vander`.
https://github.com/pytorch/pytorch/blob/630ef6c7118ed069f7c9d4f8837bd8f566ae7463/test/functorch/test_vmap_registrations.py#L182

Post the change, you can verify it locally by `python test/functorch/test_vmap_registrations.py`

**Comment by lezcano** [comment]
the failures are relevant. Please run locally the tests in  `test/functorch` with `-vk vander` before updating the PR to make sure all the relevant tests are passing.

File: test/functorch/test_vmap.py [hunk 3]
```diff
@@ -4085,7 +4084,7 @@ def f(e_):
         xfail('linalg.norm'),  # can accept vector inputs
         xfail('linalg.norm', 'subgradients_at_zero'),  # can accept vector inputs
         skip('linalg.multi_dot'),  # accepts list of tensor inputs, has its own special test
-        xfail('linalg.vander'),
+        xfail('linalg.vander'), # cannot accept dimensions equal to or less than one
```
Comment (Chillee) [inline test/functorch/test_vmap.py:4087]: Should this be failing?

Comment (jazzysoggy) [reply test/functorch/test_vmap.py:4087]: Maybe I misinterpreted the way linalg_vander works, but reading the code, it looks like it checks for whether the size of the input Tensor is greater than 1

Comment (lezcano) [reply test/functorch/test_vmap.py:4087]: The comment should probably be `# can accept vector inputs`, as in the ones above it.

**Comment by jazzysoggy** [comment]
Can I ignore the current errors seeing as they do not to the actual testing, which passed, but rather the uploading of the test results and missing a "typing extension"

**Review by lezcano — APPROVED** [review]
Right, thank you!

**Pull request closed without merging**